### PR TITLE
Add new configuration Parameter

### LINKED ISSUE #1589: Provide a way to pass a whole configuration as parameter
## Problem

As a user of GitHub CodeQL in Actions I would like to be able to pass some configuration when running my workflow, for example:
- pass a query filter from or environment variable or input parameter

Today only a subset of the available configuration/parameters is available in the action.

## Solution

Add a new parameter, for example called `configuration` that will take the same format as a `config-file` but can be used directly in the action.

The idea is to support the following examples:

- Complete on configuration
  ```yaml
  - uses: github/codeql-action/init@v2
    with:
      languages: ${{ matrix.language }}
      configuration: |
        disable-default-queries: true
        queries:
          - uses: security-extended
          - uses: security-and-quality
        query-filters:
          - include:
        tags: /cwe-020/
  ```

- Actions Variable
  ```yaml
  - uses: github/codeql-action/init@v2
    with:
      languages: ${{ matrix.language }}
      configuration: |
            ${{vars.CODEQL_CONF}}    
  ```

- Input Parameter
    ```yaml
      - uses: tgrall/codeql-action/init@query-filter
      with:
        languages: ${{ matrix.language }}
        configuration: |
          disable-default-queries: true
          queries:
            - uses: security-extended
            - uses: security-and-quality
          query-filters:
            - include:
                tags: /${{ github.event.inputs.codeql-include-tags }}/ 
   ```

### PR DESCRIPTION
**Pull request by tgrall**
- Write test to check it is read from configuration
- Update documentation

### Merge / deployment checklist

- [x] Confirm this change is backwards compatible with existing workflows.
- [x] Confirm the [readme](https://github.com/github/codeql-action/blob/main/README.md) has been updated if necessary.
- [x] Confirm the [changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md) has been updated if necessary.
- [ ] I may have to work on some exception management when the `configuration` parameter is not well formatted. 

----

close #1589 1589

### CODE REVIEW

**Comment by tgrall** [comment]
@henrymercer @alonahlobina : I have renamed the action parameter to `config` to match the `config-file` one

File: README.md [hunk 1]
```diff
@@ -135,6 +135,62 @@ By default, this will override any queries specified in a config file. If you wi
     queries: +<local-or-remote-query>,<another-query>
 ```
 
+### Configuration
```
Comment (henrymercer) [inline README.md:138]:
```suggestion
### Configuration via `config` input
```

**Review by henrymercer — COMMENTED** [review]
Thanks for your contribution!  Some suggestions to docs and code.

**Comment by tgrall** [comment]
@henrymercer 
- thanks for the review, I have applied your changes.
- I see many issues in the workflows, that were not present before, not sure what is the issue
- I was not able to do my usual manual e2e testing, as my fork cannot get the CLI 2.12.6  (it was not the case before this release)

File: README.md [hunk 2]
```diff
@@ -135,6 +135,44 @@ By default, this will override any queries specified in a config file. If you wi
...
+
+  ```yaml
+  - uses: github/codeql-action/init@v2
+    with:
+      languages: ${{ matrix.language }}
+      config: ${{ vars.CODEQL_CONF }}
+  ```
+
+
+  where `vars.CODEQL_CONF` references an [Actions configuration variable](https://docs.github.com/en/actions/learn-github-actions/variables#defining-configuration-variables-for-multiple-workflows).
```
Comment (henrymercer) [inline README.md:174]:
~~~suggestion
### Configuration via `config` input

You can alternatively configure CodeQL using the `config` input to the `init` Action. The value of this input must be a YAML string that follows the configuration file format documented at "[Using a custom configuration file](https://help.github.com/en/github/finding-security-vulnerabilities-and-errors-in-your-code/configuring-code-scanning#using-a-custom-configuration-file)."

#### Example configuration

```yaml
- uses: github/codeql-action/init@v2
  with:
    languages: ${{ matrix.language }}
    config: |
      disable-default-queries: true
      queries:
        - uses: security-extended
        - uses: security-and-quality
      query-filters:
        - include:
      tags: /cwe-020/
```

#### Sharing configuration across multiple repositories
  
You can use Actions or environment variables to share configuration across multiple repositories and to modify configuration without needing to edit the workflow file.  In the following example, `vars.CODEQL_CONF` is an [Actions configuration variable](https://docs.github.com/en/actions/learn-github-actions/variables#defining-configuration-variables-for-multiple-workflows):

```yaml
- uses: github/codeql-action/init@v2
  with:
    languages: ${{ matrix.language }}
    config: ${{ vars.CODEQL_CONF }}
```

~~~

File: src/config-utils.test.ts [hunk 3]
```diff
@@ -913,6 +925,89 @@ test("Queries in workflow file can be added to the set of queries without overri
   });
 });
 
+test("Queries can be specified in configuration, same as file", async (t) => {
```
Comment (henrymercer) [inline src/config-utils.test.ts:928]:
```suggestion
test("Queries can be specified using config input", async (t) => {
```

File: src/config-utils.test.ts [hunk 4]
```diff
@@ -913,6 +925,89 @@ test("Queries in workflow file can be added to the set of queries without overri
...
+  return await util.withTmpDir(async (tmpDir) => {
+    const inputFileContents = `
```
Comment (henrymercer) [inline src/config-utils.test.ts:930]: I think it'd be clearer to name this `configInput`.

File: src/config-utils.test.ts [hunk 5]
```diff
@@ -913,6 +925,89 @@ test("Queries in workflow file can be added to the set of queries without overri
...
+      name: my config
+      queries:
+        - uses: ./foo
```
Comment (henrymercer) [inline src/config-utils.test.ts:933]: Minor: It'd be clearer to extract `foo` into a variable so we can see more easily why we're creating the `foo` directory and referencing `foo` later on.

File: src/config-utils.test.ts [hunk 6]
```diff
@@ -913,6 +925,89 @@ test("Queries in workflow file can be added to the set of queries without overri
...
+      undefined,
+      undefined,
+      undefined,
+      undefined,
+      inputFileContents,
+      false,
+      false,
+      "",
+      "",
+      { owner: "github", repo: "example " },
```
Comment (henrymercer) [inline src/config-utils.test.ts:975]: Copied and pasted, but there shouldn't be a space here.  Please could you correct this throughout the file?

```suggestion
      { owner: "github", repo: "example" },
```

File: src/config-utils.ts [hunk 7]
```diff
@@ -1705,6 +1706,13 @@ export async function initConfig(
 ): Promise<Config> {
   let config: Config;
 
+  // if configInput is set, it takes precedence over configFile
```
Comment (henrymercer) [inline src/config-utils.ts:1709]: It might be helpful to log a warning if both `configInput` and `configFile` are set.

**Review by henrymercer — COMMENTED** [review]
Thanks — this looks good, just some minor comments.  Your PR checks failed due to hitting API rate limits.  They should hopefully pass on a retry.  If not, we can try to find a workaround.

**Review by henrymercer — COMMENTED** [review]
Thanks for addressing my comments.  Could you merge in `main` and rebuild so we can run the PR checks?

**Comment by jeffwidman** [comment]
I assume this will also support the `paths-ignore` key?

Over in :dependabot: we'd like to [configure CodeQL to ignore two files](https://github.com/dependabot/dependabot-core/pull/7126), and seemed like overkill to have to create the indirection of a separate config file just to specify those two files...

**Comment by henrymercer** [comment]
@jeffwidman 👋 Yes, all configuration keys, including `paths-ignore`, will be supported.

**Comment by aeisenberg** [comment]
What's the status of this PR? Are we just waiting for a merge with main?

**Comment by henrymercer** [comment]
If the PR checks pass, I'm happy with the PR.  @tgrall Would you mind merging in main and rebuilding the Action so we can run the checks and merge?

**Pull request closed without merging**